pkg/commands/update/versioner.go

```diff
@@ -101,91 +81,83 @@ func (g GitHub) LatestVersion(ctx context.Context) (semver.Version, error) {
 	}
 
 	_, err = io.Copy(dst, rc)
 	if err != nil {
-		return filename, fmt.Errorf("error downloading release asset: %w", err)
+		return "", fmt.Errorf("error downloading release asset: %w", err)
 	}
 
 	if err := dst.Close(); err != nil {
-		return filename, fmt.Errorf("error closing release asset file: %w", err)
+		return "", fmt.Errorf("error closing release asset file: %w", err)
 	}
 
-	assetFile := dst.Name()
+	tempFile := dst.Name()
 
-	// If we need to perform an extraction, create a temporary directory and do it there,
-	// then copy the extracted file back to the temporary one so we just track one file.
+	// TODO: We might need to also account for Window users by also checking for
+	// the .zip extension that goreleaser generates:
+	// https://github.com/fastly/cli/blob/26588cfd2d00d18643bac5cc18242b2d5ee84b34/.goreleaser.yml#L51
+	//
+	// Ideally the formats would be the same, but if that's not possible then
+	// we can look to use a genericised method such as
+	// https://pkg.go.dev/github.com/mholt/archiver#Extract for handling the
+	// extraction of a binary from the asset file instead of using the current
+	// archiver.NewTarGz().Extract() method.
 	if strings.HasSuffix(g.releaseAsset, ".tar.gz") {
-		dir, err := os.MkdirTemp(tmp, "extract")
-		if err != nil {
-			return filename, fmt.Errorf("error creating temp extraction directory: %w", err)
-		}
-		defer os.RemoveAll(dir)
-		if err := archiver.NewTarGz().Extract(assetFile, g.binary, dir); err != nil {
-			return filename, fmt.Errorf("error extracting binary: %w", err)
-		}
-		if err := os.Rename(filepath.Join(dir, g.binary), assetFile); err != nil {
-			return filename, fmt.Errorf("error renaming binary: %w", err)
+		if err := archiver.NewTarGz().Extract(tempFile, g.binary, dir); err != nil {
+			return "", fmt.Errorf("error extracting binary: %w", err)
 		}
-	}
-
-	// TODO: This is racy without the os.CreateTemp() call but does not look like it's used.
-	// Can this be removed and let the Download's caller rename the binary?
-	if g.local != "" {
-		newName := filepath.Join(tmp, g.local)
-		if err := os.Rename(assetFile, newName); err != nil {
-			return filename, fmt.Errorf("error renaming binary: %w", err)
-		}
-		assetFile = newName
+		tempFile = filepath.Join(dir, g.binary)
 	}
 
 	// G302 (CWE-276): Expect file permissions to be 0600 or less
 	// gosec flagged this:
 	// Disabling as the file was not executable without it and we need all users
 	// to be able to execute the binary.
 	/* #nosec */
-	err = os.Chmod(assetFile, 0777)
+	err = os.Chmod(tempFile, 0777)
+	if err != nil {
+		return "", err
+	}
+
+	dst, err = os.CreateTemp("", g.binary)
 	if err != nil {
-		return filename, err
+		return "", fmt.Errorf("error creating temp file: %w", err)
+	}
+
+	assetFile := dst.Name()
+
+	if err := os.Rename(tempFile, assetFile); err != nil {
+		os.Remove(assetFile)
+		return "", fmt.Errorf("error renaming temp file: %w", err)
 	}
```

Comment:
Why do the extra work of moving the file to PWD vs. just returning the asset file directly from the temp dir?

Reply:
Because it makes cleaning up on error easier without shifting the responsibility to the callee.

Reply:
Sorry, I'm probably dense, but I don't get it ☹️ The initial file created by `os.Create(filepath.Join(dir, g.releaseAsset))` will already be unique per invocation, right? How does moving it to PWD make it easier to clean up or move more responsibility to the caller? 

If you want to remove it on failure, you could do something like

```go
func Download(...) (_ string, err error) {
    f, err := os.Create(filepath.Join(dir, g.releaseAsset))
    if err != nil { ... }

    defer func(tmpfile string) { 
        if err != nil { os.Remove(tmpfile) }
    }(f.Name())

    ...
```

Reply:
It would be unique per invocation but multiple instances can run in parallel.

Reply:
Just to make sure we're on the same page, `Download()` creates a temporary directory (`dir`) inside $TMPDIR to download and uncompress the asset if needed. It then change the permissions and moves the file to a random file in $TMPDIR, returning it and removing `dir`.

Was your original question why the last part (moving the file to a random file in $TMPDIR) takes place or I misunderstood?

Reply:
> Was your original question why the last part (moving the file to a random file in $TMPDIR) takes place or I misunderstood?

Yeah, basically:

> Download() creates a temporary directory (dir) inside $TMPDIR to download and uncompress the asset if needed.

Because `dir` is created by os.MkdirTemp, it will be unique every time, right?

> It then change the permissions and moves the file to a random file in $TMPDIR, returning it and removing dir.

Given the above, isn't this unnecessary?

Reply:
I might be missing something but I don't see how I return the file within `dir` and remove the directory and its content at the same time when returning from `Download()`.

Reply:
Also, for the record, I will be closing this PR in favour of #399 and another PR yet to be opened.

Reply:
> I don't see how I can return the file within dir and remove the directory and its content at the same time when returning from Download().

The file itself is handed off to the caller and becomes their responsibility, and unless there is another requirement lurking somewhere, it doesn't matter what directory it's in. The temporary directory itself doesn't represent a resource that necessarily needs to be cleaned up by you, the OS is ultimately able to do that.

Reply:
I think it's pretty bad etiquette to create files in /tmp (or it's equivalent) and don't remove them. YMMV.
Either way, that's a behaviour I'd like to have and/or keep.

Reply:
It's your call, of course, but I'll observe that (a) this use case is exactly what OS-level temp dirs are for, and (b) you can reasonably assume the tmpdir is writable, but it's possible the current dir is not.